File: hello-go/src/server/handler/websocket_message_handler.go
package handler

import (
	"encoding/json"
	"fmt"
	"sync"

	"github.com/feuyeux/hello-mmap/hello-go/src/logger"
	"github.com/feuyeux/hello-mmap/hello-go/src/server/memory"
	"github.com/gorilla/websocket"
)

// WebSocketMessageHandler handles WebSocket message processing
type WebSocketMessageHandler struct {
	streamManager *memory.StreamManager
	memoryPool    *memory.MemoryPoolManager
	clients       map[*websocket.Conn]string
	clientsMutex  *sync.RWMutex
}

// NewWebSocketMessageHandler creates a new message handler
func NewWebSocketMessageHandler(streamMgr *memory.StreamManager, memPool *memory.MemoryPoolManager, clients map[*websocket.Conn]string, mutex *sync.RWMutex) *WebSocketMessageHandler {
	return &WebSocketMessageHandler{
		streamManager: streamMgr,
		memoryPool:    memPool,
		clients:       clients,
		clientsMutex:  mutex,
	}
}

// HandleTextMessage handles text (JSON) messages
func (h *WebSocketMessageHandler) HandleTextMessage(conn *websocket.Conn, message []byte) {
	var data WebSocketMessage
	if err := json.Unmarshal(message, &data); err != nil {
		logger.Debug(fmt.Sprintf("Invalid JSON message: %v", err))
		h.sendError(conn, "Invalid JSON format")
		return
	}

	msgType := data.Type
	if msgType == "" {
		h.sendError(conn, "Missing message type")
		return
	}

	switch msgType {
	case "START":
		h.handleStart(conn, &data)
	case "STOP":
		h.handleStop(conn, &data)
	case "GET":
		h.handleGet(conn, &data)
	default:
		logger.Debug(fmt.Sprintf("Unknown message type: %s", msgType))
		h.sendError(conn, fmt.Sprintf("Unknown message type: %s", msgType))
	}
}

// HandleBinaryMessage handles binary audio data
func (h *WebSocketMessageHandler) HandleBinaryMessage(conn *websocket.Conn, data []byte, streamID string) {
	if streamID == "" {
		logger.Debug("Received binary data but no active stream for client")
		return
	}

	logger.Debug(fmt.Sprintf("Received %d bytes of binary data for stream %s", len(data), streamID))

	// Write to stream
	h.streamManager.WriteChunk(streamID, data)
}

// handleStart handles START message (create new stream)
func (h *WebSocketMessageHandler) handleStart(conn *websocket.Conn, data *WebSocketMessage) {
	streamID := data.StreamId
	if streamID == "" {
		h.sendError(conn, "Missing streamId")
		return
	}

	// Create stream
	if h.streamManager.CreateStream(streamID) {
		// Register this client with the stream
		h.clientsMutex.Lock()
		h.clients[conn] = streamID
		h.clientsMutex.Unlock()

		response := NewStartedMessage(streamID, "Stream started successfully")
		h.sendJSON(conn, response)
		logger.Debug(fmt.Sprintf("Stream started: %s", streamID))
	} else {
		h.sendError(conn, fmt.Sprintf("Failed to create stream: %s", streamID))
	}
}

// handleStop handles STOP message (finalize stream)
func (h *WebSocketMessageHandler) handleStop(conn *websocket.Conn, data *WebSocketMessage) {
	streamID := data.StreamId
	if streamID == "" {
		h.sendError(conn, "Missing streamId")
		return
	}

	// Finalize stream
	if h.streamManager.FinalizeStream(streamID) {
		response := NewStoppedMessage(streamID, "Stream finalized successfully")
		h.sendJSON(conn, response)
		logger.Debug(fmt.Sprintf("Stream finalized: %s", streamID))

		// Unregister stream from client
		h.clientsMutex.Lock()
		h.clients[conn] = ""
		h.clientsMutex.Unlock()
	} else {
		h.sendError(conn, fmt.Sprintf("Failed to finalize stream: %s", streamID))
	}
}

// handleGet handles GET message (read stream data)
func (h *WebSocketMessageHandler) handleGet(conn *websocket.Conn, data *WebSocketMessage) {
	streamID := data.StreamId
	if streamID == "" {
		h.sendError(conn, "Missing streamId")
		return
	}

	offset := int64(0)
	if data.Offset != nil {
		offset = *data.Offset
	}

	length := 65536
	if data.Length != nil {
		length = *data.Length
	}

	// Read data from stream
	chunkData := h.streamManager.ReadChunk(streamID, offset, length)

	if len(chunkData) > 0 {
		// Send binary data
		if err := conn.WriteMessage(websocket.BinaryMessage, chunkData); err != nil {
			logger.Error(fmt.Sprintf("Error sending binary data: %v", err))
		}
		logger.Debug(fmt.Sprintf("Sent %d bytes for stream %s at offset %d", len(chunkData), streamID, offset))
	} else {
		h.sendError(conn, fmt.Sprintf("Failed to read from stream: %s", streamID))
	}
}

// sendJSON sends a JSON message to the client
func (h *WebSocketMessageHandler) sendJSON(conn *websocket.Conn, data *WebSocketMessage) {
	message, err := json.Marshal(data)
	if err != nil {
		logger.Debug(fmt.Sprintf("Error marshaling JSON: %v", err))
		return
	}

	if err := conn.WriteMessage(websocket.TextMessage, message); err != nil {
		logger.Debug(fmt.Sprintf("Error sending message: %v", err))
	}
}

// sendError sends an error message to the client
func (h *WebSocketMessageHandler) sendError(conn *websocket.Conn, message string) {
	response := NewErrorMessage(message)
	h.sendJSON(conn, response)
	logger.Debug(fmt.Sprintf("Sent error to client: %s", message))
}

File: hello-rust/src/server/handler/websocket_message_handler.rs
// WebSocket message handler for processing client messages.
// Handles START, STOP, and GET message types.

use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::HashMap;
use std::sync::{Arc, Mutex};

use crate::server::memory::{MemoryPoolManager, StreamManager};
use tungstenite::protocol::Message as WsMessage;
use tungstenite::{Bytes, Utf8Bytes, WebSocket};

/// WebSocket message types
#[derive(Debug, Serialize, Deserialize)]
pub struct ControlMessage {
    #[serde(rename = "type")]
    pub msg_type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stream_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub offset: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub length: Option<usize>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,
}

pub struct WebSocketMessageHandler;

impl WebSocketMessageHandler {
    /// Handle a text (JSON) control message.
    pub fn handle_text_message(
        websocket: &mut WebSocket<std::net::TcpStream>,
        clients: &Arc<Mutex<HashMap<usize, String>>>,
        stream_mgr: &Arc<StreamManager>,
        _mem_pool: &Arc<MemoryPoolManager>,
        client_id: usize,
        message: &str,
    ) {
        let data: Value = match serde_json::from_str(message) {
            Ok(v) => v,
            Err(e) => {
                eprintln!("Invalid JSON message: {:?}", e);
                Self::send_error(websocket, clients, client_id, "Invalid JSON format");
                return;
            }
        };

        let msg_type = data["type"].as_str().unwrap_or("");

        match msg_type {
            "START" => Self::handle_start(websocket, clients, stream_mgr, client_id, &data),
            "STOP" => Self::handle_stop(websocket, clients, stream_mgr, client_id, &data),
            "GET" => Self::handle_get(websocket, clients, stream_mgr, client_id, &data),
            _ => {
                eprintln!("Unknown message type: {}", msg_type);
                Self::send_error(
                    websocket,
                    clients,
                    client_id,
                    &format!("Unknown message type: {}", msg_type),
                );
            }
        }
    }

    /// Handle binary audio data.
    pub fn handle_binary_message(
        clients: &Arc<Mutex<HashMap<usize, String>>>,
        stream_mgr: &Arc<StreamManager>,
        client_id: usize,
        data: &[u8],
    ) {
        // Get active stream ID for this client
        let stream_id = {
            let clients = clients.lock().unwrap();
            clients.get(&client_id).cloned()
        };

        if stream_id.is_none() || stream_id.as_ref().unwrap().is_empty() {
            println!("[ERROR] Received binary data but no active stream for client");
            return;
        }

        let stream_id = stream_id.unwrap();

        // Write to stream
        stream_mgr.write_chunk(&stream_id, data);
    }

    /// Handle START message (create new stream).
    fn handle_start(
        websocket: &mut WebSocket<std::net::TcpStream>,
        clients: &Arc<Mutex<HashMap<usize, String>>>,
        stream_mgr: &Arc<StreamManager>,
        client_id: usize,
        data: &Value,
    ) {
        let stream_id = match data["streamId"].as_str() {
            Some(id) => id.to_string(),
            None => {
                Self::send_error(websocket, clients, client_id, "Missing streamId");
                return;
            }
        };

        // Create stream
        if stream_mgr.create_stream(stream_id.clone()) {
            // Register this client with the stream
            clients.lock().unwrap().insert(client_id, stream_id.clone());

            let response = ControlMessage {
                msg_type: "STARTED".to_string(), // Use uppercase to match client expectation
                stream_id: Some(stream_id.clone()),
                offset: None,
                length: None,
                message: Some("Stream created".to_string()),
            };

            Self::send_json(websocket, clients, client_id, &response);
            println!("Stream started: {}", stream_id);
        } else {
            Self::send_error(
                websocket,
                clients,
                client_id,
                &format!("Failed to create stream: {}", stream_id),
            );
        }
    }

    /// Handle STOP message (finalize stream).
    fn handle_stop(
        websocket: &mut WebSocket<std::net::TcpStream>,
        clients: &Arc<Mutex<HashMap<usize, String>>>,
        stream_mgr: &Arc<StreamManager>,
        client_id: usize,
        data: &Value,
    ) {
        let stream_id = match data["streamId"].as_str() {
            Some(id) => id.to_string(),
            None => {
                Self::send_error(websocket, clients, client_id, "Missing streamId");
                return;
            }
        };

        // Finalize stream
        if stream_mgr.finalize_stream(&stream_id) {
            let response = ControlMessage {
                msg_type: "STOPPED".to_string(), // Use uppercase to match client expectation
                stream_id: Some(stream_id.clone()),
                offset: None,
                length: None,
                message: Some("Stream finalized".to_string()),
            };

            Self::send_json(websocket, clients, client_id, &response);
            println!("Stream finalized: {}", stream_id);

            // Unregister stream from client
            clients.lock().unwrap().insert(client_id, String::new());
        } else {
            Self::send_error(
                websocket,
                clients,
                client_id,
                &format!("Failed to finalize stream: {}", stream_id),
            );
        }
    }

    /// Handle GET message (read stream data).
    fn handle_get(
        websocket: &mut WebSocket<std::net::TcpStream>,
        clients: &Arc<Mutex<HashMap<usize, String>>>,
        stream_mgr: &Arc<StreamManager>,
        client_id: usize,
        data: &Value,
    ) {
        let stream_id = match data["streamId"].as_str() {
            Some(id) => id.to_string(),
            None => {
                Self::send_error(websocket, clients, client_id, "Missing streamId");
                return;
            }
        };

        let offset = data["offset"].as_u64().unwrap_or(0);
        let length = data["length"].as_u64().unwrap_or(65536) as usize;

        // Read data from stream
        let chunk_data = stream_mgr.read_chunk(&stream_id, offset, length);

        if !chunk_data.is_empty() {
            // Send binary data via WebSocket
            match websocket.send(WsMessage::Binary(Bytes::from(chunk_data))) {
                Ok(_) => {
                    println!(
                        "Sent {} bytes for stream {} at offset {}",
                        length, stream_id, offset
                    );
                }
                Err(e) => {
                    eprintln!("Failed to send binary data: {:?}", e);
                }
            }
        } else {
            Self::send_error(
                websocket,
                clients,
                client_id,
                &format!("Failed to read from stream: {}", stream_id),
            );
        }
    }

    /// Send a JSON message to the client.
    fn send_json(
        websocket: &mut WebSocket<std::net::TcpStream>,
        _clients: &Arc<Mutex<HashMap<usize, String>>>,
        client_id: usize,
        data: &ControlMessage,
    ) {
        let json = match serde_json::to_string(data) {
            Ok(j) => j,
            Err(e) => {
                eprintln!("Error marshaling JSON: {:?}", e);
                return;
            }
        };

        // Send via WebSocket
        match websocket.send(WsMessage::Text(Utf8Bytes::from(json.as_str()))) {
            Ok(_) => {
                println!("Sending to client {}: {}", client_id, json);
            }
            Err(e) => {
                eprintln!("Failed to send message to client: {:?}", e);
            }
        }
    }

    /// Send an error message to the client.
    fn send_error(
        websocket: &mut WebSocket<std::net::TcpStream>,
        clients: &Arc<Mutex<HashMap<usize, String>>>,
        client_id: usize,
        message: &str,
    ) {
        let response = ControlMessage {
            msg_type: "ERROR".to_string(),
            stream_id: None,
            offset: None,
            length: None,
            message: Some(message.to_string()),
        };

        Self::send_json(websocket, clients, client_id, &response);
        eprintln!("Sent error to client: {}", message);
    }
}
